torch/csrc/autograd/profiler_kineto.cpp

```diff
@@ -0,0 +1,298 @@
+#include <torch/csrc/autograd/profiler_kineto.h>
+
+#include <torch/csrc/jit/frontend/tracer.h>
+#include <torch/csrc/jit/runtime/operator.h>
+
+#include <sstream>
+
+#ifndef _WIN32
+#include <pthread.h>
+#endif
+
+#ifdef USE_KINETO
+#include "libkineto.h"
+#endif
+
+namespace torch { namespace autograd { namespace profiler {
+
+#ifdef USE_KINETO
+namespace {
+// TODO: consider TLS
+uint64_t next_correlation_id() {
+  static std::atomic<uint64_t> corr_id_ {1};
+  return corr_id_++;
+}
+
+std::string shapesToStr(const std::vector<std::vector<int64_t>>& shapes);
+
+struct TORCH_API KinetoThreadLocalState : public ProfilerThreadLocalState {
+  using ProfilerThreadLocalState::ProfilerThreadLocalState;
+  virtual ~KinetoThreadLocalState() override = default;
+
+  virtual void reportClientActivity(
+      const at::RecordFunction& fn,
+      const at::ObserverContext* observer_ctx) override {
+    auto ctx = dynamic_cast<const KinetoObserverContext*>(observer_ctx);
+    TORCH_CHECK(ctx);
+    if (!ctx) {
+      return;
+    }
+    libkineto::ClientTraceActivity op;
+    op.startTime = ctx->startUs;
+    op.endTime = getTimeUs();
+    op.opType = std::string(fn.name().str());
+    op.device = 0;
+    op.threadId = ctx->startThreadId;
+    op.correlation = ctx->correlationId;
+    if (ctx->shapes && !ctx->shapes->empty()) {
+      op.inputDims = shapesToStr(*ctx->shapes);
+    }
+#ifndef _WIN32
+    op.threadId = pthread_self();
```

Comment:
@gdankel can we use platform independent threads?

Reply:
this setting is for kineto, the above is platform independent we use in pt; agree that it could be less confusing

Reply:
would be nice to have a kineto's wrapper, if we want to support windows